Modernize and some additions

pmhahn opened the pull request:
I was looking for exactly your tool; I spent some time to improve it as my patchset consists of ~200 patches and the matrix layout was just using too much space.
Also I needed an output format to feed into `tsort`, which I added.

I spent extra time to update your code to modern Python using "f-strings" (Python 3.6+) and "Walrus operator" (Python 3.8+) and "type annotations" (Python 3.6+).

I also changed the matrix code to use "Unicode box characters".

If you like please take the changes; if they are too much I can also create a reduced patch-set for you.

Thank you for providing your took and making it open-source.

pmhahn committed:
ci: Add .gitignore

pmhahn committed:
doc: Document naming

pmhahn committed:
style: white space changes

pmhahn committed:
refactor: split parse_args from main

pmhahn committed:
refactor: Remove class inheriting from object (UP004)

no longer needed with Python 3

<https://docs.astral.sh/ruff/rules/useless-object-inheritance/>

pmhahn committed:
style: Rename UnidiffParse{Exception -> Error} (N818)

<https://docs.astral.sh/ruff/rules/error-suffix-on-exception-name/>

pmhahn committed:
style: Rename {Change -> Line}

Also used to store unchanged context lines.

pmhahn committed:
refactor: Change LINE_TYPE into Enum

pmhahn committed:
refactor: Remove unused variables

pmhahn committed:
fix: Iterate patch lines only once

`./patchdeps.py --git REV` was broken as `GitRev.get_diff()` returns a
`list`, which can be walked multiple times:
`parse_diff()` starts from the beginning and delegated to _parse_hunk()`
when the header is detected, but passed on again the list, so
`_parse_hunk()` restarted from the beginning again and failed as the
first line of the patch starts with `diff…`.

pmhahn committed:
refactor: header parsing

Move conversion from `str` to `int` into `parse_diff()`.

pmhahn committed:
style: Convert to dict literal

pmhahn committed:
style: Rename `to_parse` to `{source,target}_todo`

pmhahn committed:
refactor: Convert to f-string

pmhahn committed:
refactor: Remove (re)setting `current_file`

In case of a missing source header you now get an early "AttributeError"
instead of a delayed "TypeError".

pmhahn committed:
refactor[parser]: parse_diff

Use if-elif instead of using continue.
Use walrus operator to both match and check regular expression.
Drop useless comments.

pmhahn committed:
doc: Add PEP-484 type hints

Rename second `f` to `path` as they have different types.

pmhahn committed:
refactor[parser]: Remove unneeded default values

pmhahn committed:
style: sort imports (I001)

<https://docs.astral.sh/ruff/rules/unsorted-imports/>

pmhahn committed:
feat: Implement `__repr__` to support debugging

Show useful information in `pdb`.

pmhahn committed:
refactor[parser]: Remove prefix `depends-` from arguments

too much to type

pmhahn committed:
refactor: Convert `Bunch` into an Enum

pmhahn committed:
refactor: Convert `map()` to generator expression (C417)

<https://docs.astral.sh/ruff/rules/unnecessary-map/>

pmhahn committed:
style: Drop unneeded open mode (UP015)

Reading is the default.

<https://docs.astral.sh/ruff/rules/redundant-open-modes/>

pmhahn committed:
style: Put double-quotes in single-quotes

to get rid of backslash escaping.

pmhahn committed:
refactor: Modernize list concatenation (RUF005)

<https://docs.astral.sh/ruff/rules/collection-literal-concatenation/>

pmhahn committed:
style: Use `bytes.decode()` instead of `str(encoding=…)`

pmhahn committed:
refactor: feed data via stdin to xdot

pmhahn committed:
refactor: Remove stray line

pmhahn committed:
doc: Update some comments and texts

pmhahn committed:
style: Convert to `… not in …` (E713)

<https://docs.astral.sh/ruff/rules/not-in-test/>

pmhahn committed:
fix: Comment out broken `dottooltip`

Assigns to singleton `Depend`!

pmhahn committed:
style: ´print_depends`

Use walrus operator.

pmhahn committed:
feat: Add `tsort` output

`./patchdeps.py … --tsort | tsort`

See <man:tsort(1)>

pmhahn committed:
style: swap cases to get rid of `elif` (RET505)

<https://docs.astral.sh/ruff/rules/superfluous-else-return/>

pmhahn committed:
style: join long line

pmhahn committed:
style: Simplify offset for new files

pmhahn committed:
refactor: Directly work with set-of-patches for proximity

No need to go via a number.

pmhahn committed:
refactor[matrix]: matrix printing

Use Unicode line characters and remove duplicate code

pmhahn committed:
refactor: Replace Changeset.number by `enumerate()`

pmhahn committed:
refactor[matrix]: Print blank before column

pmhahn committed:
feat[matrix]: Print more compact matrix

Only assign a column if a patch is actually referenced by a later patch.

matthijskooijman commented on patchdeps.py:
```diff
@@ -471,12 +471,12 @@ def analyze_hunk(self, depends, patch, hunk):
     def print_blame(self):
         print("{}:".format(self.fname))
-        next_line = None
+        next_line = 0
```
I do not really like this change (commit " refactor: Make next_line an int "). I had intentionally used `None` to mean "no next line" instead of using 0, since it is more explicit and does not need to assume anything about the range of line numbers in line state (though this would actually require the check a few lines down to be `next_line is not None` instead of just `next_line`, I now realize).

Maybe drop this commit? AFAICS it is not needed for other changes?

matthijskooijman commented on patchdeps.py:
```diff
@@ -471,12 +471,12 @@ def analyze_hunk(self, depends, patch, hunk):
                 continue
 
-            if next_line and line_state.lineno != next_line:
+            if line_state.lineno != next_line > 0:
```
This line looks very weird and incorrect to me. I suspect that it is actually correct and it uses a generalization of the `1 < a < 10` syntax but with different operators, but that does not make for readable code IMHO. One more reason to drop this commit?

matthijskooijman commented on patchdeps.py:
```diff
@@ -172,7 +172,7 @@ def depends_dot(args, patches, depends):
+    for i, p in enumerate(patches):
         label = dot_escape_string(str(p))
         label = "\\n".join(textwrap.wrap(label, 25))
         res += """{} [label="{}"]\n""".format(p.number, label)
```
Commit " refactor: Replace Changeset.number by enumerate() " leaves one reference to `p.number` here (and one more a few lines down). They are removed in a future commit, but those changes should be moved here.

matthijskooijman commented on patchdeps.py:
```diff
@@ -135,11 +135,11 @@ def print_depends_matrix(patches, depends):
+    for j, p in enumerate(patches):
         fill, corner = ("─", "┘") if p in has_deps else (" ", " ")
-        line = str(p)[:80] + "  " + fill * (84 - len(line) + p.number * 2) + corner + " "
+        line = str(p)[:80] + "  " + fill * (84 - len(line) + jr * 2) + corner + " "
```
Typo in commit " refactor: Replace Changeset.number by enumerate()" `jr` -> `j` (this code is refactored again later, so this error does not show up in the total change).

matthijskooijman commented on patchdeps.py:
```diff
@@ -118,20 +118,17 @@ def get_changesets(args):
-                    print("  %s (%s)" % (dep, desc))
-                else:
-                    print("  %s" % dep)
+                print(f"  {dep}{f' ({desc})' if desc else ''}")
```
In commit " refactor: Convert to f-string ", these nested f-strings are very hard to read IMHO (took me a while to figure out what was going on here), I prefer the original code with if else (which is more verbose than is nice, but is easier to read).

matthijskooijman commented on patchdeps.py:
```diff
@@ -113,8 +124,8 @@ def print_depends(patches, depends):
         for dep in patches:
-            if dep in depends[p]:
-                desc = getattr(depends[p][dep], 'desc', None)
+            if dependency := dependencies.get(dep):
```
In commit " refactor: Convert Bunch into an Enum ", this adds a reference to `dependencies`, which is not introduced until commit " style: ´print_depends` "

matthijskooijman reviewed (changes requested):
Awesome to hear my one-off tool (I haven't really used it since I wrote it) has been useful to you. And thanks for your improvements, I really like your PR. The changes are good changes and you've clearly invested time to submit them as clean and small changes, which made this easy to review :-)

There are a few commits that still have some unrelated changes in them. That's not a problem (or at least not worth the trouble fixing), except where commits introduce broken code (because the code depends on something introduced later for example). I've flagged these cases with a comment where I noticed them.

Also, commits " refactor: Convert to f-string " and " style: white space changes " also introduce some type annotations that I think would be easy to move to the later commit that is about that. Would be good if you could make that change.

I also left some other small remarks inline, but again, I'm really happy with the changes. If you can make some further corrections as per comments, I'll be happy to merge the PR.

matthijskooijman commented:
Oh, and some commit message typos:
 - " refactor[parser]; Remove unneeded default values " and " refactor: header parsing

Move conversion from string to into `parse_diff()`." should have : instead of ; I think?
 - " refactor: header parsing", second line "Move conversion from string to into `parse_diff()`." missing "int" (or extra "to").
 - " feat: Implement __repr__ to support debugging", second line "Shot useful information in `pdb`." shot -> show
 - " fix: Comment out broken dottooltip ", second line "Assigns to singelton Depend!" singelton -> singleton
 - " refactor: Remove (re)setting current_file ", second line "In case of a missing source header you not get an early "AttributeError"
instead of a delayed "TypeError".", not -> now

pmhahn commented:
Thank you for your review :clap: I hopefully fixed everything. Ready for a 2nd round.

pmhahn commented:
I have to change something first; please give me a day or so.

pmhahn commented:
> I have to change something first; please give me a day or so.

I was asked to use a different mail address for submitting; code was not changed.
Now everything is ready from my point-of-view.

matthijskooijman committed:
doc: Update README with new output format

This also changes the git hashes used in the example to point to commits
actually in the mainline kernel (previous example probably used hashes
from some non-final patchset before it was merged).

matthijskooijman committed:
fix: Remove trailing space in list output

matthijskooijman committed:
feat: Print warning when parse_diff produced no hunks

When the diff cannot be parsed (for example when git is configured to
always produce colored output), it produces no hunks and all patches
appear independent, producing confusing output.

This adds a warning to at least tell the user that something is wrong.

matthijskooijman committed:
fix: Run git diff with --no-color

This keeps patchdeps working even when the user has configured git to
always produce colored output.

matthijskooijman committed:
fix: Remove unintentional delimeters from tsort output

The `tsort` commands expect lines with two tokens separated by
whitespace (space, tab or newline). When using --git, or patch filenames
with whitespace, the tsort output would have unintended delimeters and
tsort would not work as intended.

This simply replaces these delimeters by underscores to ensure tsort
output is correct.

matthijskooijman reviewed (approved):
Awesome, your changes look good! I've also taken the code for a test drive, which worked well (after I fixed an issue with my git being configured to produce colored diff output...). 

I've taken the liberty of adding a few extra commits to the PR with some small extra things, it would be nice if you could review those :-)

While testing the tsort output, I did start to wonder what the point of it actually is? It seems like it finds an ordering where each depending patch comes after the depended-on patch, but doesn't the input patch list (necessarily) already satisfy this condition? The entire premise of this tool is that the patches are presented in-order (without missing patches), otherwise the line tracking is going to be off (or dependencies might be reversed). I guess tsort can find a *different* total ordering than the input ordering that is also valid, but it will (AFAIU) be arbitrary. Or maybe I do not understand the tsort tool well enough, that could also be the case.

pmhahn commented:
> I've taken the liberty of adding a few extra commits to the PR with some small extra things, it would be nice if you could review those :-)

I will take a look.

> While testing the tsort output, I did start to wonder what the point of it actually is? It seems like it finds an ordering where each depending patch comes after the depended-on patch, but doesn't the input patch list (necessarily) already satisfy this condition? The entire premise of this tool is that the patches are presented in-order (without missing patches), otherwise the line tracking is going to be off (or dependencies might be reversed). I guess tsort can find a _different_ total ordering than the input ordering that is also valid, but it will (AFAIU) be arbitrary. Or maybe I do not understand the tsort tool well enough, that could also be the case.

Yes: I use `tsort` to find an alternative ordering for a large patch-set. In my case I have ~200 patches for the Linux kernel, which touch many (unrelated) things. Currently they are sorted by "time", but I want to re-group them by "topic" while honoring their chunk-dependencies as much as possible. For example: if one file if changed by a single patch, I can freely apply it anytime: first or in the middle or last – ignoring any functional dependency.

pmhahn commented on patchdeps.py:
```diff
@@ -149,7 +149,10 @@ def print_depends_tsort(patches: list[Changeset], depends: dict[Changeset, dict[
             for dep in patches:
                 if dep in dependencies:
-                    print(f"{dep}\t{p}")
+                    def no_delim(x):
```
Good find; thank you for tackling this. Only some small nits:

1. Move the function definition out of the `for`-loop
2. Please also add type-hints here; `mypy` / `pytype` / … will complain otherwise.
3. Maybe better use `str.translate` instead of a chain of `str.replace()?
4. Rename the function to start with a verb?
5. Code-formaters like `ruff` will probably complain, if there is no empty line before and after the (nested) function.

```diff
+
+    def replace_tr_delim(cs: Changeset) -> str:
+        # Tsort source has: #define DELIM " \t\n"
+        return str(cs).translate(str.maketrans(" \t\n", "___"))
+
     for p in patches:
         if dependencies := depends[p]:
             for dep in patches:
                 if dep in dependencies:
-                    def no_delim(x):
-                        # Tsort source has: #define DELIM " \t\n"
-                        return str(x).replace(' ', '_').replace('\t', '_').replace('\n', '_')
-                     print(f"{no_delim(dep)}\t{no_delim(p)}")
+                     print(f"{replace_tr_delim(dep)}\t{replace_tr_delim(p)}")

pmhahn commented:
Looks good except one things related to `tsort` which you might want to fix. Thank you for your review and time.

matthijskooijman commented:
Thanks, very useful feedback :-)

I've pushed an amended commit that fixes these issues. It's slightly different from your suggestion since I started fixing things and only realized you had a full proposal at the end when I was halfway ;-p


>    Please also add type-hints here; mypy / pytype / … will complain otherwise.
Type hinting is not engrained in my habits yet, thanks.

>    Maybe better use str.translate instead of a chain of `str.replace()?
Excellent, I was already unhappy with this code, but I had not realized `str.translate()` existed. Nice!

>    Code-formaters like ruff will probably complain, if there is no empty line before and after the (nested) function.

Right, I added one after, but not before since I think that does not look nice (and my flake8 linter is ok with that).

With these changes, I'll go ahead and merge the PR. Thanks again!
